### PR TITLE
Unstructured grid dataset

### PR DESCRIPTION
**Pull request by marc-flex**
@dbochkov-flexcompute I have added the generalization of `UnstructuredGridDataset` in this other PR. In case we don't want to move forward with these changes we can potentially merge #2123 so that we don't block @daquinteroflex

### CODE REVIEW

File: tidy3d/components/heat_charge/monitor_data.py [hunk 1]
```diff
@@ -181,6 +179,22 @@ def warn_no_data(cls, val, values):
...
+        """Issue error if incorrect data type is used"""
+
+        mnt = values.get("monitor")
+
+        if isinstance(val, TetrahedralGridDataset) or isinstance(val, TriangularGridDataset):
+            if isinstance(val.values, IndexVoltageDataArray):
+                raise ValueError(
+                    f"Monitor {mnt} of type 'TemperatureMonitor' cannot be associated with data arrays "
+                    "of type 'IndexVoltageDataArray'."
+                )
```
Comment (dbochkov-flexcompute) [inline tidy3d/components/heat_charge/monitor_data.py]: since we are likely to add others indexed data arrays, probably easier to switch here to something like:
```
if not isinstance(val.values, IndexDataArray):
    raise ValueError(
        f"Unstructured grid dataset given in 'TemperatureData.temperature' must contain only one field as 'IndexedDataArray'."
    )
```

File: tidy3d/components/data/unstructured/base.py [hunk 2]
```diff
@@ -255,6 +255,8 @@ def _values_coords_dict(self):
         """Non-spatial dimensions are corresponding coordinate values of stored data."""
         coord_dict = {dim: self.values.coords[dim].data for dim in self.values.dims}
         _ = coord_dict.pop("index")
+        if "voltage" in coord_dict.keys():
+            _ = coord_dict.pop("voltage")
```
Comment (dbochkov-flexcompute) [inline tidy3d/components/data/unstructured/base.py]: the purpose of `_values_coords_dict` is to return all dimensions other than `index`. Why did you need to remove `voltage` too?

File: tidy3d/components/heat_charge/monitor_data.py [hunk 3]
```diff
@@ -251,6 +265,22 @@ def warn_no_data(cls, val, values):
...
+        """Issue error if incorrect data type is used"""
+
+        mnt = values.get("monitor")
+
+        if isinstance(val, TetrahedralGridDataset) or isinstance(val, TriangularGridDataset):
+            if isinstance(val.values, IndexVoltageDataArray):
+                raise ValueError(
+                    f"Monitor {mnt} of type 'VoltageMonitor' cannot be associated with data arrays "
+                    "of type 'IndexVoltageDataArray'."
+                )
```
Comment (dbochkov-flexcompute) [inline tidy3d/components/heat_charge/monitor_data.py]: same here

File: tidy3d/components/simulation.py [hunk 4]
```diff
@@ -4792,35 +4789,35 @@ def perturbed_mediums_copy(
...
-                new_values = IndexedDataArray(
-                    np.array(data.values.data).flatten(), coords=dict(index=data.values.index.data)
-                )
-                if num_voltages > 1:
-                    raise ValueError(
-                        f"The value entered for '{carrier}' contains field values associated to "
-                        f"{num_voltages} voltage values. Please select one before calling this function. "
-                        "This can be done with 'electron_data.sel(voltage=1)'"
+            if isinstance(data, TriangularGridDataset) or isinstance(data, TetrahedralGridDataset):
+                if "voltage" in data.values.dims:
```
Comment (dbochkov-flexcompute) [inline tidy3d/components/simulation.py]: for generality, we could probably switch this to 
```
if data._num_fields > 1:
```

File: tidy3d/components/data/unstructured/base.py (hunk 2 above)
Comment (marc-flex) [reply tidy3d/components/data/unstructured/base.py]: Yeah, I wasn't sure about this. I'll remove that

File: tidy3d/components/heat_charge/monitor_data.py (hunk 1 above)
Comment (marc-flex) [reply tidy3d/components/heat_charge/monitor_data.py]: What happens if we have a `SpatialDataArray`? This would fail right? Since it doesn't have a `values` field?

File: tidy3d/components/simulation.py (hunk 4 above)
Comment (marc-flex) [reply tidy3d/components/simulation.py]: Not sure I like this since we're later checking for the additional fields. In this case `voltage`.

**Review by daquinteroflex — APPROVED** [review]
Looks good! I'll deal with the ruff issue within the refactor PR if that sounds alright.

**Comment by daquinteroflex** [comment]
@marc-flex I've been able to run tests after fixing the ruff issue and I think a tests patch needs to be made re this PR
 https://github.com/flexcompute/tidy3d/actions/runs/12447074050/job/34750177242
 
Ideally it'd be done onto https://github.com/flexcompute/tidy3d/pull/2137 than `marc/dd-devsim` because the rebase will be hell otherwise.
